### ISSUE TITLE
fix(dart/transform) Merge setter registrations with the same name

### ISSUE THREAD

**Issue by kegluneq** [angular/angular#886]
Currently we generate one setter registration per `bind` value. For example, this annotation:

    'bind': {'text': 'tool-tip'}

attached to two classes, `A` and `B` would generate

    registerSetters({'text': (A obj, String value) => obj.text = value})
    registerSetters({'text': (B obj, String value) => obj.text = value})

Instead we need to generate a single setter registration using a common parent class of `A` and `B`. This would be tough to do without resolving in our transformer, which we would like to avoid. As a first approximation, we can omit the parameter type (`A` and `B` above) which will be interpreted as `dynamic`.

To do this, we need global knowledge of the setters which would be registered.
1. Modify the `bind_generator` transformer, its primary inputs are the `.ngDeps.dart` files generated by the app entry points.
2. Parse these files, gathering information on the setters to be generated.
3. Also parse all `.ngDeps.dart` files imported by the primary inputs.
4. Where there are multiple setters for a single property, remove type information from the registered setter.
5. Output a `.ngSetters.dart` file with a `setupReflection` method that registers all setters.
6. Generate code to import and call `setupReflection` during app setup.

[1]: https://www.dartlang.org/tools/pub/transformers/aggregate.html

**Comment by kegluneq** [comment]
After spending some more time thinking about this, the method above will not work because it will no longer be safe to use `import`s from the `.dart` file in a `.ng_deps.dart` file if we need to generate all setters into a single file.

I haven't spent much time thinking about this, but since the `reflector` passed into `setupReflection` is untyped, perhaps it would be easiest to generate a `ReflectorWrapper` that gathers the setters from all `.ng_deps.dart` files and registers the appropriate setter.

**Comment by mhevery** [comment]
@kegluneq  Can this be closed?

**Comment by kegluneq** [comment]
TL;DR: Yes

While this issue is broader than the fix in 15376a6d243740c73cf90f55525d1710cdd156f5 (that addressed duplicates in the same file, this is attempting to remove duplicates throughout the app), I think the goal of removing all duplicates is counter to our plan of "do not resolve the world". In addition, "duplicates" now exist because we're accessing the same named field on different classes, so when we generate change detector classes they will no longer look like duplicates, so I think it's safe to call this closed.
